projects/output-and-error-handling/main.go

```diff
@@ -0,0 +1,96 @@
+package main
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"strconv"
+	"time"
+)
+
+type RequestError struct {
+	SleepTime time.Duration
+	ShouldRetry     bool
+	Cause       error
+}
+func main() {
+	c := http.Client{Timeout: time.Duration(1) * time.Second}
+
+	for i := 0; i < 3; i++ {
+		resp, err := c.Get("http://localhost:8080")
+
+		if err != nil {
+			fmt.Fprint(os.Stderr, "cannot get weather at the moment. Please try again later\n")
+			os.Exit(2)
+		}
+		defer resp.Body.Close()
+		
+		body, reqErr := parseResponse(*resp)
+
+		if reqErr.Cause != nil {
+			fmt.Fprintln(os.Stderr, reqErr.Cause)
+
+			if reqErr.ShouldRetry {
+				sleep(reqErr.SleepTime)
+				continue
+			} else {
+				os.Exit(28)
+			}
+		} else {
+			fmt.Fprintln(os.Stdout, body)
+			break
+		}
+	}
+}
+
+func parseResponse(resp http.Response) (reqBody string, reqError RequestError) {
+
+	switch resp.StatusCode {
+	case 429:
+		retryTime := resp.Header.Get("Retry-After")
+		formattedRetryTime, err := ParseRetryValue(retryTime)
+
+		if err != nil {
+			// cannot determine how long to sleep for, give up
+			reqError.Cause = errors.New("cannot get weather at the moment try again later")
+			return
+		}
+		if formattedRetryTime > 5 {
+			// sleep time is long give up
+			reqError.Cause = errors.New("server is busy at the moment try again later")
+
+		} else {
+			reqError.Cause = errors.New("server is busy, retrying in few seconds")
+			reqError.SleepTime = time.Duration(formattedRetryTime)
+			reqError.ShouldRetry = true
+		}
+	default:
+		body, err := io.ReadAll(resp.Body)
+		if err != nil {
+			reqError.Cause = err
```

Comment:
Everywhere else you're setting `reqError.Cause` to a quite understandable designed-for-humans string, whereas here you're just re-using whatever error was returned from a library.

There are advantages to using designed-for-humans strings which hide maybe-confusing information, and there are advantages to preserving underlying errors which maybe help in debugging.

It seems a bit weird that this function sometimes does one and sometimes does the other. What do you think?